Objects/odictobject.c

```diff
@@ -1682,6 +1707,24 @@ odictiter_nextkey(odictiterobject *di)
     return key;
 }
 
+static PyObject *
+odictiter_nextkey(odictiterobject *di)
+{
+    if (di->di_odict == NULL)
+        return NULL;
+    if (di->di_current == NULL) {
+        Py_CLEAR(di->di_odict);
+        return NULL;
+    }
```

Comment:
These are modified by `odictiter_nextkey_lock_held` so it seems like the checks would need to be in side the critical section.